### ISSUE TITLE
Issue in simulate / fitted for mgcv::gam with binomial n/k

### ISSUE THREAD

**Issue by florianhartig** [florianhartig/DHARMa#12]
The output of the fitted function in mgcv::gam is not named, in contrast to lm / glm - this minor difference creates a problem with the simulate functions, see demonstration here

https://github.com/florianhartig/DHARMa/blob/master/Code/DHARMaPackageSupport/mgcv/GAMProblems.md

Since DHARMa 0.1.3, DHARMa overwrites the fitted function by fitted.gam in https://github.com/florianhartig/DHARMa/blob/master/DHARMa/R/compatibility.R

I'm creating this ticket to keep in mind that this is a hack that is not optimal - ideally, this would be fixed in mgcv, and once it is this should be removed again.

**Comment by pmarchand1** [comment]
Is this related to this issue? simulateResiduals fails when run with the 2-column version of binomial regression in gam, e.g.

`library(MASS)
library(mgcv)
library(DHARMa)

snails_glm <- glm(cbind(Deaths, N - Deaths) ~ Exposure, data = snails,
                  family = binomial)
plot(simulateResiduals(snails_glm))

snails_gam <- gam(cbind(Deaths, N - Deaths) ~ Exposure, data = snails,
                  family = binomial)
plot(simulateResiduals(snails_gam))  # index out of bounds error`

**Comment by florianhartig** [comment]
Thanks for the hint.

I'm not sure if it's the same problem, but I think this issue has also been around for a while. 

I have put this on the list of issues for 0.3.4, we'll have a look at that!

**Comment by florianhartig** [comment]
OK, in the course of wrapping up I have finally found the time to look at the k/n problem. The problem seems to be rooted deeply in the gam function, which is that simulations with mgcv::gam simply don't work if you provide the k/n binomial with a matrix response. 

Doing the same via weights seems to be working though. This is in the gam folder in code/packageSupport

```
library(MASS)
library(mgcv)
library(DHARMa)

snails_gam1 <- gam(cbind(Deaths, N - Deaths) ~ Exposure, data = snails, family = binomial)
plot(simulateResiduals(snails_gam1)) # index out of bounds error

snails_gam2 <- gam(Deaths/N~ Exposure, data = snails, family = binomial, weights = N)
plot(simulateResiduals(snails_gam2)) # works

simulate(snails_gam1) # doesn't work 
simulate(snails_gam2) # works
```

**Comment by florianhartig** [comment]
I think we have to clean this all up properly. I suspect that Matteo has found some ways around these issues in mgcvViz, should have a look there. For the moment, I will push this to 0.3.5

**Comment by florianhartig** [comment]
Here another issue with gam binomial submitted by Carsten Dormann

```
library(mgcv)
library(effects)
library(DHARMa)

fm <- gam(survived ~ s(age), family=binomial, data=TitanicSurvival)
sim <- simulateResiduals(fm, re.fit=T)

# hingegen funktioniert:
fm <- gam(as.numeric(survived)-1 ~ s(age), family=binomial, data=TitanicSurvival)
sim <- simulateResiduals(fm, re.fit=T)
```

**Comment by Istalan** [comment]
`TitanicSurvival$survived` is a factor with "yes" and "no", while `as.numeric(survived)-1` gives nice boolean/numeric values like 0,1.

The cause of the error seems to be, that the `simulate` function doesn't work with the factor and returns empty vectors without throwing an error itself:

```
fm <- gam(survived ~ s(age), family=binomial, data=TitanicSurvival)
# empty
simulate(fm, nsim = 2)

fm2 <- gam(as.numeric(survived)-1 ~ s(age), family=binomial, data=TitanicSurvival)
# filled appropriately
simulate(fm2, nsim = 2)

#also works
fm3 <- gam(survived == "yes" ~ s(age), family=binomial, data=TitanicSurvival)
simulate(fm3, nsim = 2)
```

So the issue lies with whatever gam uses to simulate.

**Comment by florianhartig** [comment]
Yes, I know, but this is an old Problem with mgcv::gam - it returns a very weird unnamed simulate data.frame when simulating from factors - I had already solved this but now it re-emerged with the recent change to introducing wrappers for all models, which (as I realised now) removed my previous fix for this issue. I'm currently thinking about how to fix this, I can do a quick hack or move to a more stable solution for mgcv (e.g. via using mgcvViz), which would also have a few other advantages.

**Comment by Istalan** [comment]
Oh, i'm sorry. I should have read the entire issue. I know nothing about the differences between mgcv and mgcvViz and the only wrapper i could think of would be something like this:
```
fm_new <- getRefit(fm, getObservedResponse(fm))
simulate(fm_new, nsim = 2)
```
Refitting is obviously not a good solution and it doesn't work with the k/n problem.

**Comment by florianhartig** [comment]
Done with https://github.com/florianhartig/DHARMa/pull/308, I leave this open for a final check!

**Comment by florianhartig** [comment]
OK, I can confirm that all these problems seem to be solved as long as simulations are create with mgcvViz, will close this.

**Comment by florianhartig** [comment]
As a note to the future: this problems are solved with DHARMa 0.4.5, as long as mgcViz is loaded together with mgcv!
